### DISCUSSION TITLE
What did you intend for getting event data to a state?

### DISCUSSION THREAD

**Discussion by jnz86** [kiishor/UML-State-Machine-in-C#39] [Q&A] [answered]
Your event is an integer. What was your intention if the event/signal came with some data?

For example:
```
event = event_new_connection
new_connection_id = 7
new_connection_type = 5
```

Did you intend that these would be members of the machine structure? 

Did you consider adding a pointer next to event integer as `pcontext` or something similar?

**Comment by kiishor** [comment, marked as answer]
Hello,

Yes, event is an unsigned integer in this state machine framework. It is a member of state machine structure as shown below,

```C
//! Abstract state machine structure
struct state_machine_t
{
   uint32_t Event;          //!< Pending Event for state machine
   const state_t* State;    //!< State of state machine.
};
```

Here the intention is, user shall inherit `state_machine_t` structure as shown below,

```C
// Derived state machine
struct user_state_machine
{
  state_machine_t;    // Base state machine. Must be the first member of user derived state machine.

  // User specific state machine members
  uint32_t data1;
  uint32_t data2;
  uint32_t data3;
};
```
This framework doesn't mix event and the data. handling of data shall be part of user specific state machine members in the `user_state_machine` structure. There are several ways of passing data to state machine.
- Write directly to state machine members e.g. data1, data2, ...
- state machine provides API/ function that accepts data from user.

For example, in [toaster_oven][1] state machine provides several APIs. These API can be extended to accept additional data as arguments.
I've modified the start_oven() to to accept toastTime in the example.

```C
void start_oven(oven_t* const pOven, uint32_t toastTime)
{
  pOven->Set_Time = toastTime;
  pOven->Machine.Event = EN_START;
}
```
> Make sure that you set data before setting an event.

[1]: https://github.com/kiishor/UML-State-Machine-in-C/blob/master/demo/toaster_oven/src/toaster_oven.h
